File: book_catalog.cpp
# book_catalog.cpp
/**
 * 📚 Book Catalog – Organize Your Library by Genre (C++ Edition)
 * Features: add, edit, delete, search, genre stats, persistence
 * Uses only STL, no external libraries.
 */

#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <map>
#include <set>
#include <algorithm>
#include <sstream>
#include <ctime>
#include <random>
#include <filesystem>
#include <cctype>
#include <limits>

#ifdef _WIN32
#include <windows.h>
#endif

// ─── Colors ──────────────────────────────────────────────────────────────────

#ifdef _WIN32
HANDLE hConsole;
void setColor(int color) { SetConsoleTextAttribute(hConsole, color); }
#define RESET_COLOR setColor(7)
#define COLOR_RED setColor(12)
#define COLOR_GREEN setColor(10)
#define COLOR_YELLOW setColor(14)
#define COLOR_BLUE setColor(9)
#define COLOR_MAGENTA setColor(13)
#define COLOR_CYAN setColor(11)
#define COLOR_BRIGHT setColor(15)
#define COLOR_DIM setColor(8)
#else
#define RESET_COLOR std::cout << "\x1b[0m"
#define COLOR_RED std::cout << "\x1b[31m"
#define COLOR_GREEN std::cout << "\x1b[32m"
#define COLOR_YELLOW std::cout << "\x1b[33m"
#define COLOR_BLUE std::cout << "\x1b[34m"
#define COLOR_MAGENTA std::cout << "\x1b[35m"
#define COLOR_CYAN std::cout << "\x1b[36m"
#define COLOR_BRIGHT std::cout << "\x1b[1m"
#define COLOR_DIM std::cout << "\x1b[2m"
#endif

#define C(str, color) color << str << RESET_COLOR

// ─── Helpers ─────────────────────────────────────────────────────────────────

std::string trim(const std::string& s) {
    auto start = s.find_first_not_of(" \t\n\r");
    if (start == std::string::npos) return "";
    auto end = s.find_last_not_of(" \t\n\r");
    return s.substr(start, end - start + 1);
}

std::string toLower(std::string s) {
    std::transform(s.begin(), s.end(), s.begin(), ::tolower);
    return s;
}

std::string generate_uuid() {
    static std::random_device rd;
    static std::mt19937 gen(rd());
    static std::uniform_int_distribution<> dis(0, 15);
    static std::uniform_int_distribution<> dis2(8, 11);
    std::stringstream ss;
    ss << std::hex;
    for (int i = 0; i < 8; i++) ss << dis(gen);
    ss << "-";
    for (int i = 0; i < 4; i++) ss << dis(gen);
    ss << "-4";
    for (int i = 0; i < 3; i++) ss << dis(gen);
    ss << "-";
    ss << dis2(gen);
    for (int i = 0; i < 3; i++) ss << dis(gen);
    ss << "-";
    for (int i = 0; i < 12; i++) ss << dis(gen);
    return ss.str();
}

std::string get_home_dir() {
#ifdef _WIN32
    const char* h = std::getenv("USERPROFILE");
#else
    const char* h = std::getenv("HOME");
#endif
    return h ? std::string(h) : ".";
}

// ─── Data Model ─────────────────────────────────────────────────────────────

struct Book {
    std::string id;
    std::string title;
    std::string author;
    std::string genre;
    int year;
    std::string isbn;
};

struct CatalogData {
    std::vector<Book> books;
};

// ─── JSON (simplified) ─────────────────────────────────────────────────────

std::string escape_json(const std::string& s) {
    std::string out;
    for (char c : s) {
        if (c == '"') out += "\\\"";
        else if (c == '\\') out += "\\\\";
        else if (c == '\n') out += "\\n";
        else out += c;
    }
    return out;
}

std::string serialize_books(const std::vector<Book>& books) {
    std::ostringstream json;
    json << "{\n  \"books\": [\n";
    for (size_t i = 0; i < books.size(); ++i) {
        const auto& b = books[i];
        json << "    {\n";
        json << "      \"id\": \"" << escape_json(b.id) << "\",\n";
        json << "      \"title\": \"" << escape_json(b.title) << "\",\n";
        json << "      \"author\": \"" << escape_json(b.author) << "\",\n";
        json << "      \"genre\": \"" << escape_json(b.genre) << "\",\n";
        json << "      \"year\": " << b.year << ",\n";
        json << "      \"isbn\": \"" << escape_json(b.isbn) << "\"\n";
        json << "    }";
        if (i + 1 < books.size()) json << ",";
        json << "\n";
    }
    json << "  ]\n}";
    return json.str();
}

bool deserialize_books(const std::string& json_str, std::vector<Book>& books) {
    books.clear();
    // Simple manual parser: find book objects
    size_t pos = json_str.find("\"books\":");
    if (pos == std::string::npos) return false;
    size_t start_arr = json_str.find("[", pos);
    if (start_arr == std::string::npos) return false;
    size_t end_arr = json_str.rfind("]");
    if (end_arr == std::string::npos || end_arr <= start_arr) return false;
    std::string arr = json_str.substr(start_arr + 1, end_arr - start_arr - 1);
    // Split by "{" and "}"
    size_t brace_start = arr.find("{");
    while (brace_start != std::string::npos) {
        size_t brace_end = arr.find("}", brace_start);
        if (brace_end == std::string::npos) break;
        std::string obj = arr.substr(brace_start, brace_end - brace_start + 1);
        // Parse fields
        Book b;
        auto extract = [&](const std::string& key) -> std::string {
            size_t p = obj.find("\"" + key + "\":");
            if (p == std::string::npos) return "";
            p = obj.find(":", p) + 1;
            while (p < obj.length() && (obj[p] == ' ' || obj[p] == '\n' || obj[p] == '\r')) p++;
            if (obj[p] == '"') {
                p++;
                size_t e = obj.find("\"", p);
                if (e == std::string::npos) return "";
                return obj.substr(p, e - p);
            } else {
                size_t e = obj.find_first_of(",}\n\r", p);
                if (e == std::string::npos) return "";
                return obj.substr(p, e - p);
            }
        };
        b.id = extract("id");
        b.title = extract("title");
        b.author = extract("author");
        b.genre = extract("genre");
        std::string year_str = extract("year");
        if (!year_str.empty()) b.year = std::stoi(year_str);
        else b.year = 0;
        b.isbn = extract("isbn");
        if (!b.id.empty() && !b.title.empty()) {
            books.push_back(b);
        }
        brace_start = arr.find("{", brace_end);
    }
    return true;
}

// ─── Catalog Manager ──────────────────────────────────────────────────────

class BookCatalog {
public:
    BookCatalog() {
        home = get_home_dir();
        data_dir = home + "/.book_catalog";
        std::filesystem::create_directories(data_dir);
        data_file = data_dir + "/books.json";
        load();
    }

    void load() {
        std::ifstream file(data_file);
        if (!file.is_open()) {
            books.clear();
            return;
        }
        std::stringstream buffer;
        buffer << file.rdbuf();
        file.close();
        if (!deserialize_books(buffer.str(), books)) {
            books.clear();
        }
    }

    void save() {
        std::string json = serialize_books(books);
        std::string temp = data_file + ".tmp";
        std::ofstream out(temp);
        if (out.is_open()) {
            out << json;
            out.close();
            std::filesystem::rename(temp, data_file);
        }
    }

    void addBook(const std::string& title, const std::string& author, const std::string& genre, int year, const std::string& isbn) {
        Book b;
        b.id = generate_uuid();
        b.title = title;
        b.author = author;
        b.genre = genre;
        b.year = year;
        b.isbn = isbn;
        books.push_back(b);
        save();
    }

    bool deleteBook(const std::string& id) {
        for (auto it = books.begin(); it != books.end(); ++it) {
            if (it->id == id) {
                books.erase(it);
                save();
                return true;
            }
        }
        return false;
    }

    bool updateBook(const std::string& id, const std::map<std::string, std::string>& updates) {
        for (auto& b : books) {
            if (b.id == id) {
                auto it = updates.find("title");
                if (it != updates.end()) b.title = it->second;
                it = updates.find("author");
                if (it != updates.end()) b.author = it->second;
                it = updates.find("genre");
                if (it != updates.end()) b.genre = it->second;
                it = updates.find("year");
                if (it != updates.end()) b.year = std::stoi(it->second);
                it = updates.find("isbn");
                if (it != updates.end()) b.isbn = it->second;
                save();
                return true;
            }
        }
        return false;
    }

    Book* getBook(const std::string& id) {
        for (auto& b : books) {
            if (b.id == id) return &b;
        }
        return nullptr;
    }

    std::vector<Book> search(const std::string& query, const std::string& genre,
                             const std::string& author, int yearMin, int yearMax) {
        std::vector<Book> results = books;
        if (!query.empty()) {
            std::string q = toLower(query);
            results.erase(std::remove_if(results.begin(), results.end(),
                [&](const Book& b) {
                    return !(toLower(b.title).find(q) != std::string::npos ||
                             toLower(b.author).find(q) != std::string::npos ||
                             toLower(b.genre).find(q) != std::string::npos);
                }), results.end());
        }
        if (!genre.empty()) {
            std::string g = toLower(genre);
            results.erase(std::remove_if(results.begin(), results.end(),
                [&](const Book& b) { return toLower(b.genre) != g; }), results.end());
        }
        if (!author.empty()) {
            std::string a = toLower(author);
            results.erase(std::remove_if(results.begin(), results.end(),
                [&](const Book& b) { return toLower(b.author).find(a) == std::string::npos; }), results.end());
        }
        if (yearMin > 0) {
            results.erase(std::remove_if(results.begin(), results.end(),
                [&](const Book& b) { return b.year < yearMin; }), results.end());
        }
        if (yearMax < 9999) {
            results.erase(std::remove_if(results.begin(), results.end(),
                [&](const Book& b) { return b.year > yearMax; }), results.end());
        }
        return results;
    }

    std::map<std::string, int> getGenres() const {
        std::map<std::string, int> m;
        for (const auto& b : books) {
            m[b.genre]++;
        }
        return m;
    }

    void getStats(int& total, int& authors, int& genres, int& minYear, int& maxYear) const {
        total = books.size();
        std::set<std::string> authorSet;
        for (const auto& b : books) authorSet.insert(b.author);
        authors = authorSet.size();
        genres = getGenres().size();
        if (total > 0) {
            minYear = books[0].year;
            maxYear = books[0].year;
            for (const auto& b : books) {
                if (b.year < minYear) minYear = b.year;
                if (b.year > maxYear) maxYear = b.year;
            }
        } else {
            minYear = 0;
            maxYear = 0;
        }
    }

    std::vector<Book> books;

private:
    std::string home, data_dir, data_file;
};

// ─── Main App ──────────────────────────────────────────────────────────────

class BookApp {
public:
    BookApp() : catalog() {}

    void run() {
        std::cout << "\033[2J\033[1;1H";
        std::cout << C("\n📚 Book Catalog – Organize Your Library", COLOR_BRIGHT) << C("", COLOR_CYAN) << std::endl;
        std::cout << C("Manage your books by genre!", COLOR_DIM) << std::endl;

        while (true) {
            showMenu();
            std::string choice = ask("Your choice: ");
            if (choice == "1") listBooks();
            else if (choice == "2") addBook();
            else if (choice == "3") searchBooks();
            else if (choice == "4") showStats();
            else if (choice == "5") browseByGenre();
            else if (choice == "6") editBook();
            else if (choice == "7") deleteBook();
            else if (choice == "0") {
                std::cout << C("👋 Goodbye!", COLOR_CYAN) << std::endl;
                break;
            } else {
                std::cout << C("❌ Invalid choice.", COLOR_RED) << std::endl;
            }
            if (choice != "0") {
                std::cout << "\nPress Enter to continue...";
                std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
                std::cin.get();
            }
        }
    }

private:
    BookCatalog catalog;

    std::string ask(const std::string& prompt) {
        std::cout << prompt;
        std::string line;
        std::getline(std::cin, line);
        return trim(line);
    }

    int askInt(const std::string& prompt, int def) {
        while (true) {
            std::string ans = ask(prompt);
            if (ans.empty()) return def;
            try { return std::stoi(ans); }
            catch (...) { std::cout << C("❌ Please enter a number.", COLOR_RED) << std::endl; }
        }
    }

    void showMenu() {
        std::cout << "\n" << C(std::string(50, '═'), COLOR_CYAN) << std::endl;
        std::cout << C("📚 BOOK CATALOG", COLOR_BRIGHT) << C("", COLOR_CYAN) << std::endl;
        std::cout << C(std::string(50, '═'), COLOR_CYAN) << std::endl;
        std::cout << "  Total books: " << catalog.books.size() << std::endl;
        std::cout << "  Genres: " << catalog.getGenres().size() << std::endl;
        std::cout << C(std::string(50, '═'), COLOR_CYAN) << std::endl;
        std::cout << "  1. 📖 List all books" << std::endl;
        std::cout << "  2. ➕ Add a book" << std::endl;
        std::cout << "  3. 🔍 Search books" << std::endl;
        std::cout << "  4. 📊 Statistics" << std::endl;
        std::cout << "  5. 🗂️  Browse by genre" << std::endl;
        std::cout << "  6. ✏️  Edit a book" << std::endl;
        std::cout << "  7. 🗑️  Delete a book" << std::endl;
        std::cout << "  0. 🚪 Exit" << std::endl;
        std::cout << C(std::string(50, '═'), COLOR_CYAN) << std::endl;
    }

    void listBooks() {
        if (catalog.books.empty()) {
            std::cout << C("No books found.", COLOR_YELLOW) << std::endl;
            return;
        }
        std::cout << "\n📖 BOOKS" << std::endl;
        std::cout << C(std::string(60, '─'), COLOR_DIM) << std::endl;
        for (const auto& b : catalog.books) {
            std::cout << "  " << b.title << " by " << b.author << " (" << b.year << ") – " << b.genre << " [ISBN: " << (b.isbn.empty() ? "N/A" : b.isbn) << "]" << std::endl;
        }
    }

    void addBook() {
        std::string title = ask("Title: ");
        std::string author = ask("Author: ");
        std::string genre = ask("Genre: ");
        int year = askInt("Year: ", 2000);
        std::string isbn = ask("ISBN (optional): ");
        catalog.addBook(title, author, genre, year, isbn);
        std::cout << C("✅ Book added.", COLOR_GREEN) << std::endl;
    }

    void searchBooks() {
        std::string query = ask("Search term (title/author/genre): ");
        std::string genre = ask("Filter by genre (optional): ");
        std::string author = ask("Filter by author (optional): ");
        int yearMin = askInt("Minimum year (optional): ", 0);
        int yearMax = askInt("Maximum year (optional): ", 9999);
        auto results = catalog.search(query, genre, author, yearMin, yearMax);
        if (results.empty()) {
            std::cout << C("No books match your search.", COLOR_YELLOW) << std::endl;
        } else {
            std::cout << "\n📖 SEARCH RESULTS" << std::endl;
            std::cout << C(std::string(60, '─'), COLOR_DIM) << std::endl;
            for (const auto& b : results) {
                std::cout << "  " << b.title << " by " << b.author << " (" << b.year << ") – " << b.genre << std::endl;
            }
        }
    }

    void showStats() {
        int total, authors, genres, minYear, maxYear;
        catalog.getStats(total, authors, genres, minYear, maxYear);
        auto genreMap = catalog.getGenres();
        std::cout << "\n📊 STATISTICS" << std::endl;
        std::cout << C(std::string(30, '─'), COLOR_DIM) << std::endl;
        std::cout << "  Total Books: " << total << std::endl;
        std::cout << "  Unique Authors: " << authors << std::endl;
        std::cout << "  Genres: " << genres << std::endl;
        if (total > 0) {
            std::cout << "  Oldest Book: " << minYear << std::endl;
            std::cout << "  Newest Book: " << maxYear << std::endl;
        } else {
            std::cout << "  Oldest Book: —" << std::endl;
            std::cout << "  Newest Book: —" << std::endl;
        }
        if (!genreMap.empty()) {
            std::cout << "\n📚 Books by Genre:" << std::endl;
            for (const auto& [g, count] : genreMap) {
                std::cout << "  " << g << ": " << count << std::endl;
            }
        }
    }

    void browseByGenre() {
        auto genres = catalog.getGenres();
        if (genres.empty()) {
            std::cout << C("No books yet.", COLOR_YELLOW) << std::endl;
            return;
        }
        std::vector<std::string> genreNames;
        for (const auto& [g, _] : genres) genreNames.push_back(g);
        std::sort(genreNames.begin(), genreNames.end());
        std::cout << "Select a genre:" << std::endl;
        for (size_t i = 0; i < genreNames.size(); ++i) {
            std::cout << "  " << (i+1) << ". " << genreNames[i] << " (" << genres[genreNames[i]] << ")" << std::endl;
        }
        std::string choice = ask("Number: ");
        try {
            int idx = std::stoi(choice) - 1;
            if (idx >= 0 && idx < (int)genreNames.size()) {
                std::string selected = genreNames[idx];
                auto results = catalog.search("", selected, "", 0, 9999);
                listBooks(results);
            } else {
                std::cout << C("Invalid selection.", COLOR_RED) << std::endl;
            }
        } catch (...) {
            std::cout << C("Invalid input.", COLOR_RED) << std::endl;
        }
    }

    void listBooks(const std::vector<Book>& books) {
        if (books.empty()) {
            std::cout << C("No books found.", COLOR_YELLOW) << std::endl;
            return;
        }
        std::cout << "\n📖 BOOKS" << std::endl;
        std::cout << C(std::string(60, '─'), COLOR_DIM) << std::endl;
        for (const auto& b : books) {
            std::cout << "  " << b.title << " by " << b.author << " (" << b.year << ") – " << b.genre << " [ISBN: " << (b.isbn.empty() ? "N/A" : b.isbn) << "]" << std::endl;
        }
    }

    void editBook() {
        std::string id = ask("Enter book ID to edit: ");
        Book* book = catalog.getBook(id);
        if (!book) {
            std::cout << C("Book not found.", COLOR_RED) << std::endl;
            return;
        }
        std::cout << "Editing: " << book->title << " by " << book->author << std::endl;
        std::string title = ask("Title (" + book->title + "): ");
        std::string author = ask("Author (" + book->author + "): ");
        std::string genre = ask("Genre (" + book->genre + "): ");
        std::string year_str = ask("Year (" + std::to_string(book->year) + "): ");
        std::string isbn = ask("ISBN (" + book->isbn + "): ");
        std::map<std::string, std::string> updates;
        if (!title.empty()) updates["title"] = title;
        if (!author.empty()) updates["author"] = author;
        if (!genre.empty()) updates["genre"] = genre;
        if (!year_str.empty()) updates["year"] = year_str;
        if (!isbn.empty()) updates["isbn"] = isbn;
        if (catalog.updateBook(id, updates)) {
            std::cout << C("✅ Book updated.", COLOR_GREEN) << std::endl;
        } else {
            std::cout << C("Failed to update.", COLOR_RED) << std::endl;
        }
    }

    void deleteBook() {
        std::string id = ask("Enter book ID to delete: ");
        std::string confirm = ask("Delete book " + id + "? (yes/no): ");
        if (toLower(trim(confirm)) != "yes") return;
        if (catalog.deleteBook(id)) {
            std::cout << C("🗑️  Book deleted.", COLOR_YELLOW) << std::endl;
        } else {
            std::cout << C("Book not found.", COLOR_RED) << std::endl;
        }
    }
};

int main() {
#ifdef _WIN32
    hConsole = GetStdHandle(STD_OUTPUT_HANDLE);
#endif
    try {
        BookApp app;
        app.run();
    } catch (const std::exception& e) {
        std::cerr << C("❌ Unexpected error: ", COLOR_RED) << e.what() << std::endl;
        return 1;
    }
    return 0;
}

File: BookCatalog.cs
# BookCatalog.cs
/**
 * 📚 Book Catalog – Organize Your Library by Genre (C# Edition)
 * Features: add, edit, delete, search, genre stats, persistence
 * Requires: .NET 6.0+
 */

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

class BookCatalog
{
    // ─── Colors ────────────────────────────────────────────────────────────

    private static readonly string Reset = "\u001B[0m";
    private static readonly string Bright = "\u001B[1m";
    private static readonly string Dim = "\u001B[2m";
    private static readonly string Red = "\u001B[31m";
    private static readonly string Green = "\u001B[32m";
    private static readonly string Yellow = "\u001B[33m";
    private static readonly string Blue = "\u001B[34m";
    private static readonly string Magenta = "\u001B[35m";
    private static readonly string Cyan = "\u001B[36m";

    private static string C(string text, string color) => color + text + Reset;

    // ─── Data Model ──────────────────────────────────────────────────────

    public class Book
    {
        [JsonPropertyName("id")]
        public string Id { get; set; } = "";
        [JsonPropertyName("title")]
        public string Title { get; set; } = "";
        [JsonPropertyName("author")]
        public string Author { get; set; } = "";
        [JsonPropertyName("genre")]
        public string Genre { get; set; } = "";
        [JsonPropertyName("year")]
        public int Year { get; set; }
        [JsonPropertyName("isbn")]
        public string Isbn { get; set; } = "";
    }

    // ─── Config ────────────────────────────────────────────────────────────

    private static readonly string DataDir = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
        ".book_catalog"
    );
    private static readonly string DataFile = Path.Combine(DataDir, "books.json");

    // ─── Catalog Manager ──────────────────────────────────────────────────

    private class Catalog
    {
        private readonly List<Book> books = new();

        public Catalog()
        {
            Directory.CreateDirectory(DataDir);
            Load();
        }

        private void Load()
        {
            if (!File.Exists(DataFile))
            {
                books.Clear();
                return;
            }
            try
            {
                string json = File.ReadAllText(DataFile);
                var data = JsonSerializer.Deserialize<Dictionary<string, List<Book>>>(json);
                if (data != null && data.ContainsKey("books"))
                {
                    books.Clear();
                    books.AddRange(data["books"]);
                }
            }
            catch
            {
                books.Clear();
            }
        }

        private void Save()
        {
            var data = new Dictionary<string, List<Book>> { ["books"] = books };
            string json = JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = true });
            File.WriteAllText(DataFile, json);
        }

        public Book AddBook(string title, string author, string genre, int year, string isbn)
        {
            var book = new Book
            {
                Id = Guid.NewGuid().ToString(),
                Title = title,
                Author = author,
                Genre = genre,
                Year = year,
                Isbn = isbn
            };
            books.Add(book);
            Save();
            return book;
        }

        public bool DeleteBook(string id)
        {
            int removed = books.RemoveAll(b => b.Id == id);
            if (removed > 0) { Save(); return true; }
            return false;
        }

        public bool UpdateBook(string id, Book updates)
        {
            var book = books.FirstOrDefault(b => b.Id == id);
            if (book == null) return false;
            if (!string.IsNullOrEmpty(updates.Title)) book.Title = updates.Title;
            if (!string.IsNullOrEmpty(updates.Author)) book.Author = updates.Author;
            if (!string.IsNullOrEmpty(updates.Genre)) book.Genre = updates.Genre;
            if (updates.Year > 0) book.Year = updates.Year;
            if (!string.IsNullOrEmpty(updates.Isbn)) book.Isbn = updates.Isbn;
            Save();
            return true;
        }

        public Book GetBook(string id) => books.FirstOrDefault(b => b.Id == id);

        public List<Book> Search(string query, string genre, string author, int? yearMin, int? yearMax)
        {
            var results = books.AsEnumerable();
            if (!string.IsNullOrEmpty(query))
            {
                var q = query.ToLower();
                results = results.Where(b =>
                    b.Title.ToLower().Contains(q) ||
                    b.Author.ToLower().Contains(q) ||
                    b.Genre.ToLower().Contains(q)
                );
            }
            if (!string.IsNullOrEmpty(genre))
                results = results.Where(b => b.Genre.Equals(genre, StringComparison.OrdinalIgnoreCase));
            if (!string.IsNullOrEmpty(author))
                results = results.Where(b => b.Author.ToLower().Contains(author.ToLower()));
            if (yearMin.HasValue) results = results.Where(b => b.Year >= yearMin.Value);
            if (yearMax.HasValue) results = results.Where(b => b.Year <= yearMax.Value);
            return results.ToList();
        }

        public Dictionary<string, int> GetGenres()
        {
            return books.GroupBy(b => b.Genre)
                        .ToDictionary(g => g.Key, g => g.Count());
        }

        public (int total, int authors, int genres, int? minYear, int? maxYear) GetStats()
        {
            int total = books.Count;
            int authors = books.Select(b => b.Author).Distinct().Count();
            int genres = GetGenres().Count;
            int? minYear = books.Any() ? books.Min(b => b.Year) : (int?)null;
            int? maxYear = books.Any() ? books.Max(b => b.Year) : (int?)null;
            return (total, authors, genres, minYear, maxYear);
        }
    }

    // ─── Main App ──────────────────────────────────────────────────────────

    private readonly Catalog catalog;

    public BookCatalog()
    {
        catalog = new Catalog();
    }

    private string Ask(string prompt)
    {
        Console.Write(prompt);
        return Console.ReadLine()?.Trim() ?? "";
    }

    private int AskInt(string prompt, int def)
    {
        while (true)
        {
            string ans = Ask(prompt);
            if (string.IsNullOrEmpty(ans)) return def;
            if (int.TryParse(ans, out int val)) return val;
            Console.WriteLine(C("❌ Please enter a number.", Red));
        }
    }

    private void ShowMenu()
    {
        Console.WriteLine("\n" + C(new string('═', 50), Cyan));
        Console.WriteLine(C("📚 BOOK CATALOG", Bright + Cyan));
        Console.WriteLine(C(new string('═', 50), Cyan));
        Console.WriteLine($"  Total books: {catalog.books.Count}");
        Console.WriteLine($"  Genres: {catalog.GetGenres().Count}");
        Console.WriteLine(C(new string('═', 50), Cyan));
        Console.WriteLine("  1. 📖 List all books");
        Console.WriteLine("  2. ➕ Add a book");
        Console.WriteLine("  3. 🔍 Search books");
        Console.WriteLine("  4. 📊 Statistics");
        Console.WriteLine("  5. 🗂️  Browse by genre");
        Console.WriteLine("  6. ✏️  Edit a book");
        Console.WriteLine("  7. 🗑️  Delete a book");
        Console.WriteLine("  0. 🚪 Exit");
        Console.WriteLine(C(new string('═', 50), Cyan));
    }

    private void ListBooks(List<Book> books = null)
    {
        if (books == null) books = catalog.books;
        if (!books.Any())
        {
            Console.WriteLine(C("No books found.", Yellow));
            return;
        }
        Console.WriteLine("\n📖 BOOKS");
        Console.WriteLine(C(new string('─', 60), Dim));
        foreach (var b in books)
        {
            Console.WriteLine($"  {b.Title} by {b.Author} ({b.Year}) – {b.Genre} [ISBN: {b.Isbn}]");
        }
    }

    private void AddBook()
    {
        string title = Ask("Title: ");
        string author = Ask("Author: ");
        string genre = Ask("Genre: ");
        int year = AskInt("Year: ", 2000);
        string isbn = Ask("ISBN (optional): ");
        var book = catalog.AddBook(title, author, genre, year, isbn);
        Console.WriteLine(C($"✅ Book '{book.Title}' added with ID {book.Id}", Green));
    }

    private void SearchBooks()
    {
        string query = Ask("Search term (title/author/genre): ");
        string genre = Ask("Filter by genre (optional): ");
        string author = Ask("Filter by author (optional): ");
        int yearMin = AskInt("Minimum year (optional): ", 0);
        int yearMax = AskInt("Maximum year (optional): ", 9999);
        var results = catalog.Search(
            query, genre, author,
            yearMin > 0 ? yearMin : (int?)null,
            yearMax < 9999 ? yearMax : (int?)null
        );
        if (results.Any())
            ListBooks(results);
        else
            Console.WriteLine(C("No books match your search.", Yellow));
    }

    private void ShowStats()
    {
        var (total, authors, genres, minYear, maxYear) = catalog.GetStats();
        var genreMap = catalog.GetGenres();
        Console.WriteLine("\n📊 STATISTICS");
        Console.WriteLine(C(new string('─', 30), Dim));
        Console.WriteLine($"  Total Books: {total}");
        Console.WriteLine($"  Unique Authors: {authors}");
        Console.WriteLine($"  Genres: {genres}");
        Console.WriteLine($"  Oldest Book: {minYear?.ToString() ?? "—"}");
        Console.WriteLine($"  Newest Book: {maxYear?.ToString() ?? "—"}");
        if (genreMap.Any())
        {
            Console.WriteLine("\n📚 Books by Genre:");
            foreach (var g in genreMap.OrderBy(x => x.Key))
                Console.WriteLine($"  {g.Key}: {g.Value}");
        }
    }

    private void BrowseByGenre()
    {
        var genres = catalog.GetGenres();
        if (!genres.Any())
        {
            Console.WriteLine(C("No books yet.", Yellow));
            return;
        }
        var genreNames = genres.Keys.OrderBy(x => x).ToList();
        Console.WriteLine("Select a genre:");
        for (int i = 0; i < genreNames.Count; i++)
            Console.WriteLine($"  {i+1}. {genreNames[i]} ({genres[genreNames[i]]})");
        string choice = Ask("Number: ");
        if (int.TryParse(choice, out int idx) && idx >= 1 && idx <= genreNames.Count)
        {
            string selected = genreNames[idx-1];
            var books = catalog.Search("", selected, "", null, null);
            ListBooks(books);
        }
        else
            Console.WriteLine(C("Invalid selection.", Red));
    }

    private void EditBook()
    {
        string id = Ask("Enter book ID to edit: ");
        var book = catalog.GetBook(id);
        if (book == null)
        {
            Console.WriteLine(C("Book not found.", Red));
            return;
        }
        Console.WriteLine($"Editing: {book.Title} by {book.Author}");
        string title = Ask($"Title ({book.Title}): ");
        string author = Ask($"Author ({book.Author}): ");
        string genre = Ask($"Genre ({book.Genre}): ");
        string yearStr = Ask($"Year ({book.Year}): ");
        string isbn = Ask($"ISBN ({book.Isbn}): ");
        var updates = new Book
        {
            Title = string.IsNullOrEmpty(title) ? book.Title : title,
            Author = string.IsNullOrEmpty(author) ? book.Author : author,
            Genre = string.IsNullOrEmpty(genre) ? book.Genre : genre,
            Year = string.IsNullOrEmpty(yearStr) ? book.Year : int.Parse(yearStr),
            Isbn = string.IsNullOrEmpty(isbn) ? book.Isbn : isbn
        };
        if (catalog.UpdateBook(id, updates))
            Console.WriteLine(C("✅ Book updated.", Green));
        else
            Console.WriteLine(C("Failed to update.", Red));
    }

    private void DeleteBook()
    {
        string id = Ask("Enter book ID to delete: ");
        string confirm = Ask($"Delete book {id}? (yes/no): ");
        if (!confirm.Equals("yes", StringComparison.OrdinalIgnoreCase)) return;
        if (catalog.DeleteBook(id))
            Console.WriteLine(C("🗑️  Book deleted.", Yellow));
        else
            Console.WriteLine(C("Book not found.", Red));
    }

    public void Run()
    {
        Console.Clear();
        Console.WriteLine(C("\n📚 Book Catalog – Organize Your Library", Bright + Cyan));
        Console.WriteLine(C("Manage your books by genre!", Dim));

        while (true)
        {
            ShowMenu();
            string choice = Ask("Your choice: ");
            switch (choice)
            {
                case "1": ListBooks(); break;
                case "2": AddBook(); break;
                case "3": SearchBooks(); break;
                case "4": ShowStats(); break;
                case "5": BrowseByGenre(); break;
                case "6": EditBook(); break;
                case "7": DeleteBook(); break;
                case "0":
                    Console.WriteLine(C("👋 Goodbye!", Cyan));
                    return;
                default:
                    Console.WriteLine(C("❌ Invalid choice.", Red));
                    break;
            }
            if (choice != "0")
            {
                Console.Write("\nPress Enter to continue...");
                Console.ReadLine();
            }
        }
    }

    public static void Main()
    {
        try
        {
            new BookCatalog().Run();
        }
        catch (Exception ex)
        {
            Console.WriteLine(C($"❌ Unexpected error: {ex.Message}", Red));
            Environment.Exit(1);
        }
    }
}
